src/parser.rs

```diff
@@ -0,0 +1,229 @@
+//! Provides a way to easily decode types from a `&[u8]`.
+
+use crate::{FromBytes, Ref, Unaligned};
+use core::mem::size_of;
+
+/// Parses values and types
+#[derive(Debug, Clone)]
+pub struct Parser<'a> {
+    /// The bytes that have not yet been decoded.
+    rest: &'a [u8],
+}
+
+/// This type is returned by Parser methods.
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+pub struct ParserError;

```

Comment:
This `ParserError` type exists so that users of the crate can define their own `From<ParserError>` implementations, which convert from `ParserError` to their own error types.  Because `ParserError` implements `std::error::Error`, it can also be converted to `anyhow::Error`.

Both of these are ergonomic benefits, and they would be lost if these functions returned `Option` instead.